File: tobas_setup_assistant/src/tobas_setup_assistant/setting_widgets/barometer.py
from __future__ import annotations
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ..setup_assistant import SetupAssistant
    from ..parameter_getters import ParamGetterWidget

from typing import override
from PyQt5.QtWidgets import QVBoxLayout

from ..common import SENSOR_OFFSET_DESCRIPTION
from ..parameter_getters import (
    ParamGetterWidget_SpinBox,
    ParamGetterWidget_DoubleSpinBox,
    ParamGetterWidget_Vector3d,
)
from .base_setting import BaseSettingWidget


class BarometerWidget(BaseSettingWidget):
    NAME = "Barometer"
    TITLE_TEXT = "Define Air Pressure Sensor"
    ABST_TEXT = ""  # TODO

    def __init__(self, main: SetupAssistant) -> None:
        super().__init__(main)

        self._param_rows = QVBoxLayout()
        self._rows.addLayout(self._param_rows)

        self.offset = ParamGetterWidget_Vector3d("Offset", SENSOR_OFFSET_DESCRIPTION, suffix=" m")
        self._param_rows.addWidget(self.offset)

        update_rate_description = ""
        self.update_rate = ParamGetterWidget_SpinBox(
            "Update Rate", update_rate_description, minimum=1, default=50, suffix=" Hz"
        )
        self._param_rows.addWidget(self.update_rate)

        pressure_var_description = ""
        self.pressure_var = ParamGetterWidget_DoubleSpinBox(
            "Air Pressure Variance",
            pressure_var_description,
            decimals=2,
            minimum=0.0,
            default=10.0,
            suffix=" Pa^2",
        )
        self._param_rows.addWidget(self.pressure_var)

        self._rows.addStretch()

    @override
    def update_internal_data_structures(self) -> None:
        pass

    @override
    def is_valid(self) -> bool:
        return True

    @override
    def dump_settings(self) -> dict:
        res = dict()

        for i in range(self._param_rows.count()):
            param: ParamGetterWidget = self._param_rows.itemAt(i).widget()
            res[param.name()] = param.get()

        return res

    @override
    def load_settings(self, data: dict) -> None:
        for i in range(self._param_rows.count()):
            param: ParamGetterWidget = self._param_rows.itemAt(i).widget()
            param.set(data[param.name()])

    def equipped(self) -> bool:
        return True

File: tobas_gui/tobas_setup_assistant/src/setting_widgets/barometer.cpp
#include <tobas_yaml_tools/convert/eigen.hpp>

#include "tobas_setup_assistant/setting_tabs/barometer.hpp"

namespace gui
{
namespace sa
{
BarometerWidget::BarometerWidget()
{
  offset_ = new ParamGetterWidget_Vector3d("Offset", kSensorOffsetDescription);
  offset_->setSuffix(" m");
  addWidget(offset_);

  update_rate_ = new ParamGetterWidget_SpinBox("Update Rate", "");  // TODO
  update_rate_->setMinimum(1);
  update_rate_->setValue(50);
  update_rate_->setSuffix(" Hz");
  addWidget(update_rate_);

  pressure_var_ = new ParamGetterWidget_DoubleSpinBox("Air Pressure Variance", "");  // TODO
  pressure_var_->setDecimals(2);
  pressure_var_->setMinimum(0.);
  pressure_var_->setValue(10.);
  pressure_var_->setSuffix(" Pa^2");
  addWidget(pressure_var_);

  addStretch();
}

const char* BarometerWidget::name() const
{
  return "Barometer";
}

const char* BarometerWidget::title() const
{
  return "Define Air Pressure Sensor";
}

const char* BarometerWidget::description() const
{
  return "";  // TODO
}

void BarometerWidget::onOpened()
{
  return;
}

void BarometerWidget::updateInternalDataStructures()
{
  return;
}

bool BarometerWidget::isValid()
{
  return true;
}

YAML::Node BarometerWidget::dump()
{
  YAML::Node node(YAML::NodeType::Map);

  node[offset_->name()] = offset_->getValue();
  node[update_rate_->name()] = update_rate_->getValue();
  node[pressure_var_->name()] = pressure_var_->getValue();

  return node;
}

void BarometerWidget::load(const YAML::Node& node)
{
  offset_->setValue(node[offset_->name()].as<Eigen::Vector3d>());
  update_rate_->setValue(node[update_rate_->name()].as<int>());
  pressure_var_->setValue(node[pressure_var_->name()].as<double>());
}

Eigen::Vector3d BarometerWidget::offset() const
{
  return offset_->getValue();
}

int BarometerWidget::updateRate() const
{
  return update_rate_->getValue();
}

double BarometerWidget::pressureVariance() const
{
  return pressure_var_->getValue();
}
}  // namespace sa
}  // namespace gui
